### PR TITLE
Update discussion of parameter type specifications in docstrings

### PR DESCRIPTION
**Pull request by namurphy**
This PR expands the discussion on type lines in docstrings, since it didn't have enough information before.  I'm planning to base a lot of it on the [matplotlib doc guide](https://matplotlib.org/stable/devel/documenting_mpl.html#parameter-type-descriptions) and on https://github.com/numpy/numpydoc/pull/367.  In particular, I want to add a bunch of type line examples, including for infrequent cases.

### CODE REVIEW

**Comment by namurphy** [comment]
It looks like there isn't a standard practice for multi-line type specifications: https://github.com/numpy/numpydoc/issues/87

**Comment by namurphy** [comment]
Oh, the [LSST docstring guide](https://developer.lsst.io/python/numpydoc.html) has a [section on docstring parameter types](https://developer.lsst.io/python/numpydoc.html#py-docstring-parameter-types), with some examples pretty much from numpydoc.  There are similarities and differences with matplotlib's conventions.

**Comment by namurphy** [comment]
A lot of my motivation for this PR is that the numpydoc standard doesn't cover a bunch of edge cases that come up for packages that depend on NumPy pretty often, but seemingly don't come up for NumPy itself.  I am trying to be consistent with primarily the [matplotlib style guide](https://matplotlib.org/stable/devel/documenting_mpl.html#parameter-type-descriptions) and also the [LSST standards](https://developer.lsst.io/python/numpydoc.html#py-docstring-parameter-types).

The numpydoc standard would have been a much better place to clear up edge cases, but it seems pretty difficult to change since the PR I submitted on `keyword-only` and `positional-only` has been waiting on validation functionality to be added.

My main motivations for this PR are that standardizing the type specifications (ideally across projects) will:
 - **Make documentation easier to read because of improved consistency**
 - Use links to glossary terms to make docstrings more concise and less [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
 - Set the stage for docstring autoformatters

**Comment by namurphy** [comment]
@pheuer — might you be availble to review this? I was thinking you might have insight into parameter type description edge cases that are likely to come up in `plasmapy.diagnostics`, and possibly notice things that are missing.  Thank you!

**Review by StanczakDominik — APPROVED** [review]
To me, it looks okay. Having taken a look at your numpydoc PRs, the only worry I have is extending our docs too much beyond what the numpydoc standard currently offers. I don't want to break tooling that's based on a strict (as in, validated) interpretation of the standard.

But that's probably a worry for another day.

I also wonder how this plays with xarray, which I'd like to integrate in one day. Then you don't need to track the dimension number or ordering of your array because you're operating on named dimensions anyway.

But that's also a worry for another day.

**Comment by namurphy** [comment]
> Having taken a look at your numpydoc PRs, the only worry I have is extending our docs too much beyond what the numpydoc standard currently offers. I don't want to break tooling that's based on a strict (as in, validated) interpretation of the standard.

Agreed.  Going with matplotlib's style guide when possible partially addresses that worry, but not completely.

At the moment our docstrings go beyond what's in the numpydoc standard in a bunch of places. For example, a lot of what's in [`nitpick_ignore_regex`](https://github.com/PlasmaPy/PlasmaPy/blob/60097f8b4741f59a367dab2640cd3c7360c7737f/docs/conf.py#L331) is in there for the parameter type specifications for the nitpicky doc builds.  Consistentifying docstring practices will mean we can simplify what's in `nitpick_ignore_regex` too.

> I also wonder how this plays with xarray

Interesting!  I looked at xarray's contributor guide and didn't see much about docstrings.

> But that's also a worry for another day.

That is absolutely my favorite type of worry! 😹

File: docs/contributing/doc_guide.rst [hunk 1]
```diff
@@ -949,54 +949,160 @@ Docstring guidelines
...
+^^^^^^^^^^^^^^^^^^^^^^
+
+The **parameter description** should concisely describe the meaning of
+the parameter, as well as any requirements or restrictions on allowed
+values of the parameter (including those specified by
+|validate_quantities| or |particle_input|. The parameter description
+should not repeat information already in the type specification, but may
+include type information when:
+
+- There isn't enough space in the type specification;
```
Comment (pheuer) [inline docs/contributing/doc_guide.rst:1082]: Does this mean to keep it within the line length limit?

File: docs/contributing/doc_guide.rst [hunk 2]
```diff
@@ -949,54 +949,160 @@ Docstring guidelines
...
+
+   *args : tuple, optional
+       Description of positional arguments.
+
+   **kwargs : dict, optional
+       Description of keyword arguments.
+
+.. todo::
+
+   Add a convention for multi-line type specifications.
```
Comment (pheuer) [inline docs/contributing/doc_guide.rst:1102]: This would be great...

**Review by pheuer — APPROVED** [review]
All looks good - I think examples are really helpful for this but you've included a good number of them (and more are obviously in the code...)

File: docs/contributing/doc_guide.rst (hunk 2 above)
Comment (namurphy) [reply docs/contributing/doc_guide.rst:1102]: Agreed!  I tried out some options, but have yet to find one that Sphinx renders as intended into the online documentation.  Sigh!

File: docs/contributing/doc_guide.rst (hunk 1 above)
Comment (namurphy) [reply docs/contributing/doc_guide.rst:1082]: Yes.  I'll clarify this.

**Pull request merged**